torch/csrc/autograd/python_variable.cpp

```diff
@@ -144,6 +144,32 @@ static PyObject *THPVariable_pynew(PyTypeObject *type, PyObject *args, PyObject
 }
 
 // Instantiates a subclass of torch.Tensor. Used by nn.Parameter()
+static PyObject* THPVariable_as_subclass(THPVariable* self, PyObject* args, PyObject* kwargs) {
+  HANDLE_TH_ERRORS
+  static PythonArgParser parser({
+    "as_subclass(PyObject* cls)",
+  });
+  ParsedArgs<2> parsed_args{};
+  auto r = parser.parse(args, kwargs, parsed_args);
+  PyObject* cls = r.pyobject(0);
+  if (!PyType_Check(cls)) {
+    throw TypeError("cls must be a type (got %s)", Py_TYPE(cls)->tp_name);
+  }
+  auto data = as_variable_ref(self->cdata);
+  // We set `data`'s `allow_tensor_metadata_change` to true here, because we want to
+  // allow the following use case for backward compatibility:
+  //
+  // ```python
+  // rnn = torch.nn.RNN(100, 100, 2)
+  // # The following calls `torch._cudnn_rnn_flatten_weight(rnn._flat_weights, ...)`,
+  // # which changes storage of `rnn`'s weights in-place
+  // rnn.flatten_parameters()
+  // ```
+  data.unsafeGetTensorImpl()->set_allow_tensor_metadata_change(true);
+  return THPVariable_NewWithVar((PyTypeObject*)cls, std::move(data));
```

Comment:
I think your modifications here make this unsound. `data` corresponds directly to the passed in Python object (not a new alias of it, as was the case in `THPVariable_make_subclass`, and then you call `THPVariable_NewWithVar` which has this invariant:

```
// Creates a new Python object for a Variable. The Variable must not already
// have a PyObject* associated with it.
static PyObject* THPVariable_NewWithVar(PyTypeObject* type, Variable var)
```

This invariant is violated: you already have a PyObject, the one the variable came in with.

Reply:
Would there be a way to make an alias that wouldn't violate this invariant and keep its position in the computation graph?

Reply:
Had a quick chat with @ezyang: We agreed that `.alias()` was the way to go, it'll keep the new object attached to autograd as an alias of the original, so autograd should continue to function.